### PR TITLE
[Merged by Bors] - remove blanket `Serialize + Deserialize` requirement for `Reflect` on generic types

### PR DESCRIPTION
**Pull request by jakobhellermann**
# Objective

Some generic types like `Option<T>`, `Vec<T>` and `HashMap<K, V>` implement `Reflect` when where their generic types `T`/`K`/`V` implement `Serialize + for<'de> Deserialize<'de>`.
This is so that in their `GetTypeRegistration` impl they can insert the `ReflectSerialize` and `ReflectDeserialize` type data structs.

This has the annoying side effect that if your struct contains a `Option<NonSerdeStruct>` you won't be able to derive reflect (https://github.com/bevyengine/bevy/issues/4054).

## Solution

- remove the `Serialize + Deserialize` bounds on wrapper types
  - this means that `ReflectSerialize` and `ReflectDeserialize` will no longer be inserted even for `.register::<Option<DoesImplSerde>>()`
- add `register_type_data<T, D>` shorthand for `registry.get_mut(T).insert(D::from_type<T>())`
- require users to register their specific generic types **and the serde types** separately like
```rust
        .register_type::<Option<String>>()
        .register_type_data::<Option<String>, ReflectSerialize>()
        .register_type_data::<Option<String>, ReflectDeserialize>()

```
I believe this is the best we can do for extensibility and convenience without specialization.

## Changelog

- `.register_type` for generic types like `Option<T>`, `Vec<T>`, `HashMap<K, V>` will no longer insert `ReflectSerialize` and `ReflectDeserialize` type data. Instead you need to register it separately for concrete generic types like so:
```rust
        .register_type::<Option<String>>()
        .register_type_data::<Option<String>, ReflectSerialize>()
        .register_type_data::<Option<String>, ReflectDeserialize>()
```

TODO: more docs and tweaks to the scene example to demonstrate registering generic types.

### CODE REVIEW

File: crates/bevy_app/src/app.rs [hunk 1]
```diff
@@ -867,6 +867,28 @@ impl App {
         self
     }
 
+    /// Adds the type data `D` to type `T` in the [TypeRegistry](bevy_reflect::TypeRegistry) resource
```
Comment (alice-i-cecile) [inline crates/bevy_app/src/app.rs:870]:
```suggestion
    /// Adds the type data `D` to type `T` in the [TypeRegistry](bevy_reflect::TypeRegistry) resource.
    ///
```

File: crates/bevy_app/src/app.rs [hunk 2]
```diff
@@ -857,7 +857,7 @@ impl App {
         self
     }
 
-    /// Adds the type `T` to the type registry [`Resource`].
+    /// Adds the type `T` to the [TypeRegistry](bevy_reflect::TypeRegistry) resource
```
Comment (alice-i-cecile) [inline crates/bevy_app/src/app.rs:860]:
```suggestion
    /// Adds the type `T` to the [`TypeRegistry`](bevy_reflect::TypeRegistry) resource
```
```suggestion
    /// Adds the type `T` to the [TypeRegistry](bevy_reflect::TypeRegistry) resource
```

File: crates/bevy_app/src/app.rs (hunk 1 above)
Comment (alice-i-cecile) [inline crates/bevy_app/src/app.rs:870]: I'd like to see some brief compare-and-contrast between this and its sibling to help guide users.

File: crates/bevy_reflect/src/type_registry.rs [hunk 3]
```diff
@@ -100,6 +100,22 @@ impl TypeRegistry {
             .insert(registration.type_id(), registration);
     }
 
+    /// Registers the type data `D` for type `T`
```
Comment (alice-i-cecile) [inline crates/bevy_reflect/src/type_registry.rs:103]:
```suggestion
    /// Registers the type data `D` for type `T`.
    ///
```
```suggestion
    /// Registers the type data `D` for type `T`
```

File: crates/bevy_reflect/src/type_registry.rs [hunk 4]
```diff
@@ -100,6 +100,23 @@ impl TypeRegistry {
...
+    /// # Example
+    /// ```rust
+    /// use bevy_reflect::{TypeRegistry, ReflectSerialize, ReflectDeserialize};
+    ///
+    /// let mut type_registry = TypeRegistry::default();
+    /// type_registry.register::<Option<String>>();
+    /// type_registry.register_type_data::<Option<String>, ReflectSerialize>();
+    /// type_registry.register_type_data::<Option<String>, ReflectDeserialize>();
+    /// ```
+    pub fn register_type_data<T: 'static, D: TypeData + FromType<T>>(&mut self) -> Option<()> {
```
Comment (MrGVSV) [inline crates/bevy_reflect/src/type_registry.rs:123]: Should we instead do this to ensure the given type can even be registered?
```suggestion
    pub fn register_type_data<T: GetTypeRegistration + 'static, D: TypeData + FromType<T>>(&mut self) -> Option<()> {
```

File: crates/bevy_reflect/src/type_registry.rs [hunk 5]
```diff
@@ -100,6 +104,28 @@ impl TypeRegistry {
...
     }
 
+    /// Registers the type data `D` for type `T`.
+    ///
+    /// Most of the time [`TypeRegistry::register`] can be used instead to register a type you derived [`Reflect`] for.
+    /// However, in cases where you want to add a piece of type data that was not included in the list of `#[reflect(...)]` type data in the derive,
+    /// or where the type is generic and cannot register e.g. `ReflectSerialize` unconditionally without knowing the specific type parameters,
+    /// this method can be used to insert additional type data.
+    ///
+    /// # Example
```
Comment (MrGVSV) [inline crates/bevy_reflect/src/type_registry.rs:114]: Why return `Option`? Should this return a `Result` that errors if the type is not yet registered (and gives a helper error message as to what went wrong)?

**Review by alice-i-cecile — APPROVED** [review]

**Review by PROMETHIA-27 — APPROVED** [review]
Looks good! And I've just noticed that there was a big change to TypeData, gonna have to familiarize myself with that.

File: crates/bevy_reflect/src/type_registry.rs (hunk 5 above)
Comment (makspll) [reply crates/bevy_reflect/src/type_registry.rs:114]: I would agree here. The fully explicit registering approach is better in this case imo.
when I first started working with Reflection I definitely relied on "Magic", and I wouldn't be surprised if others did too especially in this case. Panicking early with a helpful message once is much better than silent failures later, and directs the user to correct long-term usage of this API.

File: crates/bevy_reflect/src/type_registry.rs (hunk 4 above)
Comment (makspll) [reply crates/bevy_reflect/src/type_registry.rs:123]: > I don't think we need to even require Reflect for having something in the TypeRegistry

Correct me if I am wrong but you can only create a `TypeRegistration` via `TypeRegistration::of` which forces a `Reflect` bound on `T`

And as for registering something you can only do it via `TypeRegistry::register` which enforces a `GetTypeRegistration` bound or via `TypeRegistry::add_registration` for which you need a Reflect bound to create the registration.

Are there any plans to allow non reflect types in the registry in the future? If not I would say that requiring at least `Reflect` on this method would provide better guidance for users, at least with the current state of the public API .

**Review by alice-i-cecile — APPROVED** [review]
bors r+

**Pull request closed without merging**